pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,293 @@
+		masterMachineSelector:                 masterMachineSelector,
+		masterMachineLister:                   machinelistersv1beta1.NewMachineLister(masterMachineInformer.GetIndexer()),
+		configMapListerForKubeSystemNamespace: kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Lister().ConfigMaps("kube-system"),
+	}
+	return factory.New().
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		WithInformers(
+			masterMachineInformer,
+			kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Informer(),
+		).
+		ResyncEvery(1*time.Minute). // an arbitrary number, not too high not too low
+		ToController("MachineDeletionHooksController", eventRecorder.WithComponentSuffix("cluster-machine-deletion-hooks-controller"))
+}
+
+func (c *machineDeletionHooksController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	// stop if the machine API is not functional
+	if isFunctional, err := c.machineAPIChecker.IsFunctional(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	if err := c.addDeletionHookToMasterMachines(ctx); err != nil {
+		return err
+	}
+	// attempt to remove hooks from machines pending deletion and without running etcd member
+	return c.attemptToDeleteMachineDeletionHook(ctx, syncCtx.Recorder())
+}
+
+// it will only remove the hook once it has identified that the Machine resource is being deleted and a replacement member has been created.
+func (c *machineDeletionHooksController) attemptToDeleteMachineDeletionHook(ctx context.Context, recorder events.Recorder) error {
+	desiredControlPlaneReplicasCount, err := readDesiredControlPlaneReplicasCount(c.configMapListerForKubeSystemNamespace)
+	if err != nil {
+		return err
+	}
+	masterMachines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	if err != nil {
+		return err
+	}
+	// machines with the deletion hooks are the ones that host etcd members
+	memberMachines := filterMachinesWithMachineDeletionHook(masterMachines)
+
+	// if the number of master machines is > controlPlane.replicas based on data from the cache
+	if len(memberMachines) <= desiredControlPlaneReplicasCount {
+		klog.V(4).Infof("haven't found a replacement machine, the number of desired control plane replicas: %d must be greater than the current number of machines that host an etcd member: %d", desiredControlPlaneReplicasCount, len(memberMachines))
+		return nil
+	}
+
+	// if at least one member machine is pending deletion
+	memberMachinesPendingDeletion := filterMachinesPendingDeletion(memberMachines)
+	if len(memberMachinesPendingDeletion) == 0 {
+		klog.V(4).Infof("currently, we don't have a machine pending deletion")
+		return nil
+	}
+
+	// get the members, issue a live call instead of reading from the config map
+	// because we need to take into account learners
+	members, err := c.etcdClient.MemberList(ctx)
+	if err != nil {
+		return err
+	}
+	currentMemberIPListSet := sets.NewString()
```

Comment:
`currentVotingMemberIPListSet`?

Reply:
nope, the list holds both voting members and learners.
we also need to remove hooks from machines that host "non-promotable" learners.